File: hashtable.py
#!/usr/bin/python3

def safeIdx(arr, idx):
    try:
        return arr[idx]
    except IndexError:
        return None


class HashTable:

    hashes = []
    values = []

    def __init__(self):
        for i in range(128):
            self.hashes.append(-i)
            self.values.append(-i)

    def hash(self, data):
        hash = 0

        for c in data:
            chr = ord(c)
            hash = ((hash << 5) - hash) + chr
            hash |= 0  # conv to 32bit int

        return hash

    def add(self, key, value):
        h = self.hash(key)
        idx = h % len(self.hashes)

        existingId = safeIdx(self.hashes, idx)

        existing = existingId == h

        if existing:
            self.values[idx] = value
        else:
            if idx >= len(self.hashes):
                self.hashes.append(h)
                self.values.append(value)
            else:
                self.hashes[idx] = h
                self.values[idx] = value

    def get(self, key):
        h = self.hash(key)
        idx = h % len(self.hashes)

        res = None

        if self.hashes[idx] == h:
            res = safeIdx(self.values, idx)

        return res


ht = HashTable()

ht.add('foo', 'bar')
ht.add('TEST', 'HUI')

assert ht.get('foo') == 'bar'
assert ht.get('qwe') is None
assert ht.get('TEST') == 'HUI'
assert ht.get('TESTZ') is None

File: hashtable.rs
struct HashTable {
    hashes: Vec<i32>,
    values: Vec<String>,
}

impl HashTable {
    fn new(init_size: i32) -> HashTable {
        let mut hashes: Vec<i32> = Vec::new();
        let mut values: Vec<String> = Vec::new();

        for _ in 0..init_size {
            hashes.push(-1);
            values.push("".to_string()); // how to use -1?
        }

        HashTable { hashes, values }
    }

    fn hash(&mut self, data: &String) -> i32 {
        let mut h: i32 = 0;

        for c in data.chars() {
            let ascii = c as i32;
            h = ((h << 5) - h) + ascii;
            h |= 0
        }

        h
    }

    fn add(&mut self, key: String, value: String) {
        let h: i32 = self.hash(&key);
        let idx: i32 = h % self.hashes.len() as i32;

        let existing_id: i32 = self.hashes[idx as usize];

        if existing_id >= 0 {
            self.values[idx as usize] = value;
        } else {
            if idx >= self.hashes.len() as i32 {
                self.hashes.push(h);
                self.values.push(value)
            } else {
                self.hashes[idx as usize] = h;
                self.values[idx as usize] = value
            }
        }
    }

    fn get(&mut self, key: String) -> Option<String> {
        let h: i32 = self.hash(&key);
        let idx: i32 = h % self.hashes.len() as i32;

        if self.hashes[idx as usize] == h {
            return Some(self.values[idx as usize].clone());
        }

        None
    }
}

fn main() {
    let mut ht: HashTable = HashTable::new(128);

    ht.add("foo".to_string(), "bar".to_string());
    ht.add("TEST".to_string(), "HUI".to_string());

    assert!(ht.get("foo".to_string()).unwrap() == "bar".to_string());
    ht.add("foo".to_string(), "QWE".to_string());
    assert!(ht.get("foo".to_string()).unwrap() == "QWE".to_string());
    assert!(ht.get("qwe".to_string()).is_none() == true);
    assert!(ht.get("TEST".to_string()).unwrap() == "HUI".to_string());
    assert!(ht.get("TESTZ".to_string()).is_none() == true);
}
